Data race detected in unit tests

The following command
```
go test -race ./state/mpt/... -count 1
```
identifies the following data race:
```
==================
WARNING: DATA RACE
Write at 0x00c000274c60 by goroutine 11076:
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*nodeBase).MarkClean()
      /home/herbert/coding/fantom/carmen/go/state/mpt/nodes.go:379 +0x31
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*AccountNode).MarkClean()
      <autogenerated>:1 +0x17
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*writeBuffer).emptyBuffer()
      /home/herbert/coding/fantom/carmen/go/state/mpt/write_buffer.go:200 +0x57c
  github.com/Fantom-foundation/Carmen/go/state/mpt.makeWriteBuffer.func1()
      /home/herbert/coding/fantom/carmen/go/state/mpt/write_buffer.go:101 +0x1a6

Previous read at 0x00c000274c60 by goroutine 11077:
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*nodeBase).IsDirty()
      /home/herbert/coding/fantom/carmen/go/state/mpt/nodes.go:375 +0x31
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*AccountNode).IsDirty()
      <autogenerated>:1 +0x17
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).addToCacheHoldingTransferMutex()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:749 +0x110
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getSharedNode()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:560 +0xa25
  github.com/Fantom-foundation/Carmen/go/state/mpt.getAccess[go.shape.struct { github.com/Fantom-foundation/Carmen/go/state/mpt/shared.handle = github.com/Fantom-foundation/Carmen/go/state/mpt/shared.handle[github.com/Fantom-foundation/Carmen/go/state/mpt.Node] }]()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:615 +0x49
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).getWriteAccess()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:674 +0x6f
  github.com/Fantom-foundation/Carmen/go/state/mpt.(*Forest).SetAccountInfo()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:306 +0x7a
  github.com/Fantom-foundation/Carmen/go/state/mpt.testForest_WriteBufferRecoveryIsThreadSafe.func4()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest_test.go:666 +0x2d6

Goroutine 11076 (running) created at:
  github.com/Fantom-foundation/Carmen/go/state/mpt.makeWriteBuffer()
      /home/herbert/coding/fantom/carmen/go/state/mpt/write_buffer.go:85 +0x397
  github.com/Fantom-foundation/Carmen/go/state/mpt.makeForest()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest.go:291 +0xf7d
  github.com/Fantom-foundation/Carmen/go/state/mpt.testForest_WriteBufferRecoveryIsThreadSafe()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest_test.go:599 +0x6cb
  github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest_test.go:569 +0x28
  testing.tRunner()
      /home/herbert/sdk/go1.21.0/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /home/herbert/sdk/go1.21.0/src/testing/testing.go:1648 +0x44

Goroutine 11077 (running) created at:
  github.com/Fantom-foundation/Carmen/go/state/mpt.testForest_WriteBufferRecoveryIsThreadSafe()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest_test.go:659 +0xe91
  github.com/Fantom-foundation/Carmen/go/state/mpt.TestForest_WriteBufferRecoveryIsThreadSafe()
      /home/herbert/coding/fantom/carmen/go/state/mpt/forest_test.go:569 +0x28
  testing.tRunner()
      /home/herbert/sdk/go1.21.0/src/testing/testing.go:1595 +0x238
  testing.(*T).Run.func1()
      /home/herbert/sdk/go1.21.0/src/testing/testing.go:1648 +0x44
==================
--- FAIL: TestForest_WriteBufferRecoveryIsThreadSafe (0.26s)
    testing.go:1465: race detected during execution of test
FAIL
```

--- a/go/state/mpt/shared/shared.go
+++ b/go/state/mpt/shared/shared.go
@@ -264,6 +264,24 @@ func (h *WriteHandle[T]) Set(value T) {
 	h.shared.value = value
 }
 
+// AsReadHandle obtains a view on this write handle proving read access to the
+// shared value. Write access is preserved and must still be released. The
+// resulting read access handle must not be released.
+// TODO: split access permission proofs and handles
+// See https://github.com/Fantom-foundation/Carmen/issues/719
+func (h *WriteHandle[T]) AsReadHandle() ReadHandle[T] {
+	return ReadHandle[T]{h.handle}
+}
+
+// AsViewHandle obtains a view on this write handle proving view access to the
+// shared value. Write access is preserved and must still be released. The
+// resulting view access handle must not be released.
+// TODO: split access permission proofs and handles
+// See https://github.com/Fantom-foundation/Carmen/issues/719
+func (h *WriteHandle[T]) AsViewHandle() ViewHandle[T] {
+	return ViewHandle[T]{h.handle}
+}
+
 // Release abandons the access permission on the underlying shared object, allowing
 // other operations to gain access. It must be called eventually on all valid
 // instances to avoid dead-lock situations. After the handle has been released,

--- a/go/state/mpt/write_buffer.go
+++ b/go/state/mpt/write_buffer.go
@@ -190,9 +190,9 @@ func (b *writeBuffer) emptyBuffer() {
 		}
 
 		// Write a snapshot of the node to the disk.
-		handle := node.GetViewHandle()
+		handle := node.GetWriteHandle() // write access is needed to clear the dirty flag.
 		if handle.Get().IsDirty() {
-			if err := b.sink.Write(id, handle); err != nil {
+			if err := b.sink.Write(id, handle.AsViewHandle()); err != nil {
 				b.errsMutex.Lock()
 				b.errs = append(b.errs, err)
 				b.errsMutex.Unlock()

